File: Java/src/TwoSum.java
import java.util.Arrays;
import java.util.HashMap;

public class TwoSum {
    public static void main(String[] args) {
        int[] arr = {1, 2, 5, 7, 7};
        System.out.println(Arrays.toString(solve(arr, 7)));     // [1, 2]
        System.out.println(Arrays.toString(solveViaHashMap(arr, 7)));   // [1, 2]
    }

    static int[] solve(int[] nums, int target) {
//        bruteforce solution
//        O(n^2)
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                if (nums[i] + nums[j] == target) return new int[] {i, j};
            }
        }
        return new int[] {-1, -1};
    }

    static int[] solveViaHashMap(int[] nums, int target) {
//      O(n)
        HashMap<Integer, Integer> map = new HashMap<>();
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            int diff = target - nums[i];
            if (map.containsKey(diff)) return new int[] {map.get(diff), i};
            map.put(nums[i], i);
        }
        return new int[] {-1, -1};
    }
}

File: Python/TwoSum.py
def solve(nums, target):
    # bruteforce solution
    # O(n^2)
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]
    return [-1, -1]


def solveViaHashMap(nums, target):
    # O(n)
    hashmap = {}
    for i in range(len(nums)):
        diff = target - nums[i]
        if diff in hashmap:
            return [hashmap[diff], i]
        hashmap[nums[i]] = i
    return [-1, -1]


def main():
    arr = [1, 5, 2, 4]
    print(solve(arr, 10))
    print(solveViaHashMap(arr, 6))


if __name__ == '__main__':
    main()
